wikimedia/wikipedia-ios
# Refactor article reference code into several protocols, add to EditPreviewVC

**Fixes Phabricator ticket:**  https://phabricator.wikimedia.org/T253894

### Notes
* This is a bunch of refactoring, then a tiny bit of code to allow EditPreviewVC to use the shared code. Two new protocols were added (`ArticleScrolling`, `ReferenceShowing`), along with extensive protocol extensions for these and existing protocols.
* I changed `EditPreviewViewController` to no longer use storyboards. One of the protocols accessed `articleURL`. That variable is an optional `URL?` on `EditPreviewVC`, but a non-optional `URL` on `ArticleVC`. Unfortunately, they both either need to be optional or non-optional to conform to the same protocol. It turns out we were inflating `EditPreviewVC` from a storyboard, and that is the only reason `articleURL` (along with several other class variables) were optional. On top of that, the storyboard was fairly simplistic. I've removed the storyboard and recreated the VC's view in code, but that may be a controversial decision. Open to discussing whether this is the best path to accomplish this.
* The references protocols have turned into a bit of spaghetti. `WMFReferencePageViewAppearanceDelegate`, `ReferenceViewControllerDelegate`, `ReferenceBackLinksViewControllerDelegate`, and `ReferenceShowing` are distinct protocols, yet could probably be combined if we wanted. I could try to refactor them into a better package as part of this PR - let me know.
* When testing, you may sometimes find a reference that shows as blank. This is already being tracked in a separate bug: https://phabricator.wikimedia.org/T251830 This also can break jumping up and down between the "Preview of References" section and the actual references in the section preview: the app still queues up a completion (for the jump to a broken reference) which never fires - but then the next time you tap a (working) reference, its completion is queued up but the first completion is the one that actually fires.

### Test Steps
Note that you may need to clean and build to get it all working.

Confirm no regressions on both iPad and iPhone articles:
1. In an article, tap a reference link. Ensure it works as normal.
1. Tap a where there are multiple references for one sentence.
1. From the references section, tap a backlink - the article should scroll up to the reference, you should get the references chrome, and the reference number you tapped should be in the middle of the screen and highlighted. If there are multiple spots that use that reference, you should be able to navigate between them. Tapping elsewhere in the webview should close the references chrome.
1. Tap the reference again, and you should get the reference inspector that shows the reference over the article.
1. Tap the icon in the top left corner (of the mini-reference VC) to jump back down to the `Preview of References` section. The reference you were on should be highlight for a second, and then fade to normal.
1. Tap a backlink, you should be scrolled up into the article again. Tap the icon in the top left corner (of the backlink VC) to jump back down to the `Preview of References` section. The reference you were on should be highlight for a second, and then fade to normal.

Confirm references now work for EditPreview (on both iPad and iPhone):
1. Note the final bullet point above - if you tap a reference in the article which isn't actually in the `Preview of References`, it may seem broken for the remainder of your time navigating around that Edit Preview VC.
1. In an article, make any edit, tap "next" to see edit preview.
1. Tap a reference link, ensure it shows the reference.
1. Tap a where there are multiple references for one sentence.
1. From the references section, tap a backlink - the article should scroll up to the reference, you should get the references chrome, and the reference number you tapped should be in the middle of the screen and highlighted. If there are multiple spots that use that reference, you should be able to navigate between them. Tapping elsewhere in the webview should close the references chrome.
1. Tap the reference again, and you should get the reference inspector that shows the reference over the article.
1. Tap the icon in the top left corner (of the mini-reference VC) to jump back down to the `Preview of References` section. The reference you were on should be highlight for a second, and then fade to normal.
1. Tap a backlink, you should be scrolled up into the article again. Tap the icon in the top left corner (of the backlink VC) to jump back down to the `Preview of References` section. The reference you were on should be highlight for a second, and then fade to normal.
